src/components/tl/mlx5/mcast/tl_mlx5_mcast_team.c

```diff
@@ -227,20 +212,77 @@ ucc_status_t ucc_tl_mlx5_mcast_coll_setup_comm_resources(ucc_tl_mlx5_mcast_coll_
     comm->pending_recv = 0;
     comm->buf_n        = comm->params.rx_depth * 2;
 
-    ret = posix_memalign((void**) &comm->pp_buf, page_size, buf_size * comm->buf_n);
-    if (ret) {
-        tl_error(comm->ctx->lib, "posix_memalign failed");
-        return UCC_ERR_NO_MEMORY;
-    }
+    if (comm->device_mem_enabled) {
+        /* TODO add check to make sure GPUDirect is enabled
+         * lsmod | grep nv_peer */
+        CUDA_FUNC(cudaMalloc((void **)&comm->grh_buf, GRH_LENGTH * sizeof(char)));
+        if (!comm->grh_buf) {
+            tl_error(comm->ctx->lib, "cuda memcpy failed");
+            status = UCC_ERR_NO_MEMORY;
+            goto error;
+        }
 
-    memset(comm->pp_buf, 0, buf_size * comm->buf_n);
-    
-    comm->pp_mr = ibv_reg_mr(comm->ctx->pd, comm->pp_buf, buf_size * comm->buf_n,
-                             IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_LOCAL_WRITE);
-    if (!comm->pp_mr) {
-        tl_error(comm->ctx->lib, "could not register pp_buf mr, errno %d", errno);
-        status = UCC_ERR_NO_MEMORY;
-        goto error;
+        CUDA_FUNC(cudaMemset(comm->grh_buf, 0, GRH_LENGTH));
+
+        comm->grh_mr = ibv_reg_mr(comm->ctx->pd, comm->grh_buf, GRH_LENGTH,
+                                  IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_LOCAL_WRITE);
+        if (!comm->grh_mr) {
+            tl_error(comm->ctx->lib, "Could not register device memory for GRH, errno %d", errno);
+            status = UCC_ERR_NO_RESOURCE;
+            goto error;
+        }
+
+        // assuming the device page size is same as host page size
```

Comment:
how important is this assumption? Can you add a check?

Reply:
Thanks, Sam for noticing this line. This line is no longer applicable here, so I will remove it.